src/Microsoft.AspNet.Mvc.Core/Formatters/AcceptHeaderParser.cs

```diff
@@ -0,0 +1,165 @@
+{
+    public static class AcceptHeaderParser
+    {
+        private static readonly StringSegment QualityParameter = new StringSegment("q");
+
+        public static IList<MediaTypeSegmentWithQuality> ParseAcceptHeader(IList<string> acceptHeaders)
+        {
+            var parsedValues = new List<MediaTypeSegmentWithQuality>();
+            ParseAcceptHeader(acceptHeaders, parsedValues);
+
+            return parsedValues;
+        }
+
+        public static void ParseAcceptHeader(IList<string> acceptHeaders, IList<MediaTypeSegmentWithQuality> parsedValues)
+        {
+            if (acceptHeaders == null)
+            {
+                throw new ArgumentNullException(nameof(acceptHeaders));
+            }
+
+            if (parsedValues == null)
+            {
+                throw new ArgumentNullException(nameof(parsedValues));
+            }
+
+            for (var i = 0; i < acceptHeaders.Count; i++)
+            {
+                var index = 0;
+                var value = acceptHeaders[i];
+
+                while (!string.IsNullOrEmpty(value) && index < value.Length)
+                {
+                    MediaTypeSegmentWithQuality output;
+                    if (TryParseValue(value, ref index, out output))
+                    {
+                        // The entry may not contain an actual value, like " , "
+                        if (output.MediaType.HasValue)
+                        {
+                            parsedValues.Add(output);
+                        }
+                    }
+                    else
+                    {
+                        throw new FormatException(string.Format(CultureInfo.InvariantCulture, "Invalid values '{0}'.",
+                            value.Substring(index)));
+                    }
+                }
+            }
+        }
+
+        private static bool TryParseValue(string value, ref int index, out MediaTypeSegmentWithQuality parsedValue)
+        {
+            parsedValue = default(MediaTypeSegmentWithQuality);
+
+            // If multiple values are supported (i.e. list of values), then accept an empty string: The header may
+            // be added multiple times to the request/response message. E.g.
+            //  Accept: text/xml; q=1
+            //  Accept:
+            //  Accept: text/plain; q=0.2
+            if (string.IsNullOrEmpty(value) || (index == value.Length))
+            {
+                return true;
+            }
+
+            var separatorFound = false;
+            var currentIndex = GetNextNonEmptyOrWhitespaceIndex(value, index, out separatorFound);
+
+            if (currentIndex == value.Length)
+            {
+                index = currentIndex;
+                return true;
+            }
+
+            MediaTypeSegmentWithQuality result;
+            var length = GetMediaTypeWithQualityLength(value, currentIndex, out result);
+
+            if (length == 0)
+            {
+                return false;
```

Comment:
You handle 'empty' in `ParseAcceptHeader` as if this should return true. What's the deal?